include/swift/Basic/SupplementaryOutputPaths.h

```diff
@@ -21,49 +21,106 @@
   /// It is valid whenever there are any inputs.
+  ///
+  /// This binary format stores doc comments and other information about the
+  /// declarations in a module.
+  ///
+  /// \sa swift::serialize
   std::string ModuleDocOutputPath;
 
   /// The path to which we should output a Make-style dependencies file.
   /// It is valid whenever there are any inputs.
+  ///
+  /// Swift's compilation model means that Make-style dependencies aren't
+  /// well-suited to model fine-grained dependencies. See docs/Driver.md for
+  /// more information.
+  ///
+  /// \sa ReferenceDependenciesFilePath
   std::string DependenciesFilePath;
 
-  /// The path to which we should output a Swift reference dependencies file.
+  /// The path to which we should output a Swift "reference dependencies" file.
   /// It is valid whenever there are any inputs.
+  ///
+  /// "Reference dependencies" track dependencies on a more fine-grained level
+  /// than just "this file depends on that file". With Swift's "implicit
+  /// visibility" within a module, that becomes very important for any sort of
+  /// incremental build. These files are consumed by the Swift driver to decide
+  /// whether a source file needs to be recompiled during a build. See
+  /// docs/DependencyAnalysis.rst for more information.
+  ///
+  /// \sa swift::emitReferenceDependencies
+  /// \sa DependencyGraph
   std::string ReferenceDependenciesFilePath;
 
   /// Path to a file which should contain serialized diagnostics for this
   /// frontend invocation.
+  ///
+  /// This uses the same serialized diagnostics format as Clang, for tools that
+  /// want machine-parseable diagnostics. There's a bit more information on
+  /// how clients might use this in docs/Driver.md.
+  ///
+  /// \sa swift::serialized_diagnostics::createConsumer
   std::string SerializedDiagnosticsPath;
 
   /// The path to which we should output fix-its as source edits.
+  ///
+  /// This is a JSON-based format that is used by the migrator, but is not
+  /// really vetted for anything else.
+  ///
+  /// \sa swift::writeEditsInJson
   std::string FixItsOutputPath;
 
   /// The path to which we should output a loaded module trace file.
+  ///
+  /// This is a JSON-based format that encodes which modules have been loaded.
+  /// It's based on the same file-based information as the Make-style
+  /// dependencies.
+  ///
   /// It is currently only used with WMO, but could be generalized.
   std::string LoadedModuleTracePath;
 
   /// The path to which we should output a TBD file.
-  /// It is currently only used with WMO, but could be generalized.
+  ///
+  /// "TBD" stands for "text-based dylib". It's a YAML-based format that
+  /// describes the public ABI of a library, which clients can link against
+  /// without having an actual dynamic library binary.
+  ///
+  /// Only makes sense when the compiler has whole-module knowledge.
+  ///
+  /// \sa swift::writeTBDFile
   std::string TBDPath;
```

Comment:
I'll poke the Clang folks to see if there's any documentation I can link to.

Reply:
…apparently there isn't today. Oh well.

Reply:
I want to say (may not be accurate) "consumed by the driver to optimize linking" or whatever the accurate version is.

Reply:
"which clients can link against without having an actual dynamic library binary" doesn't cover that? (It's consumed by the linker.)

Reply:
Yes, the Apple linker `ld`/`ld64` has support for these text-based stubs, and, as Jordan says, it can consume them for linking.

Reply:
Yep. Whether or not that's supported is an implementation detail of a particular linker, though. I'm pretty sure Apple's ld64 and maybe LLVM's lld are the only ones.